src/Build/BackEnd/Components/ProjectCache/ProjectCacheService.cs

```diff
@@ -0,0 +1,240 @@
+        {
+            if (pluginDescriptor.PluginInstance != null)
+            {
+                return pluginDescriptor.PluginInstance;
+            }
+            if (pluginDescriptor.PluginAssemblyPath != null)
+            {
+                return GetPluginInstanceFromType(GetTypeFromAssemblyPath(pluginDescriptor.PluginAssemblyPath));
+            }
+
+            ErrorUtilities.ThrowInternalErrorUnreachable();
+
+#pragma warning disable CS8603 // Possible null reference return.
+            return null;
+#pragma warning restore CS8603 // Possible null reference return.
+        }
+
+        private static ProjectCachePluginBase GetPluginInstanceFromType(Type pluginType)
+        {
+            try
+            {
+                return (ProjectCachePluginBase) Activator.CreateInstance(pluginType);
+            }
+            catch (TargetInvocationException e) when (e.InnerException != null)
+            {
+                throw e.InnerException;
+            }
+        }
+
+        private static Type GetTypeFromAssemblyPath(string pluginAssemblyPath)
+        {
+            var assembly = LoadAssembly(pluginAssemblyPath);
+
+            var type = GetTypes<ProjectCachePluginBase>(assembly).FirstOrDefault();
+
+            ErrorUtilities.VerifyThrow(type != null, "NoProjectCachePluginFoundInAssembly", pluginAssemblyPath);
+
+            return type!;
+
+            Assembly LoadAssembly(string resolverPath)
+            {
+#if !FEATURE_ASSEMBLYLOADCONTEXT
+                return Assembly.LoadFrom(resolverPath);
+#else
+                return _loader.LoadFromPath(resolverPath);
+#endif
+            }
+
+            IEnumerable<Type> GetTypes<T>(Assembly assembly)
+            {
+                return assembly.ExportedTypes
+                    .Select(type => new {type, info = type.GetTypeInfo()})
+                    .Where(
+                        t => t.info.IsClass &&
+                             t.info.IsPublic &&
+                             !t.info.IsAbstract &&
+                             typeof(T).IsAssignableFrom(t.type))
+                    .Select(t => t.type);
+            }
+        }
+
+#if FEATURE_ASSEMBLYLOADCONTEXT
+        private static readonly CoreClrAssemblyLoader _loader = new CoreClrAssemblyLoader();
+#endif
+
+        public async Task<CacheResult> GetCacheResultAsync(BuildRequestData buildRequest)
+        {
+            // TODO: Parent these logs under the project build event so they appear nested under the project in the binlog viewer.
+            var queryDescription = $"{buildRequest.ProjectFullPath}" +
+                                   $"\n\tTargets:[{string.Join(", ", buildRequest.TargetNames)}]" +
+                                   $"\n\tGlobal Properties: {{{string.Join(",", buildRequest.GlobalProperties.Select(kvp => $"{kvp.Name}={kvp.EvaluatedValue}"))}}}";
+
+            _logger.LogMessage(
+                "\n====== Querying project cache for project " + queryDescription,
```

Comment:
It will most likely change in future PRs. I'd say it's fine to delay localization as long as it's in the experimental namespace and only likely used internally.